File: src/lib/algorithm/preprocess.ts
// Module 2: Preprocessing
// Normalizes feedback text and converts it to tokenized machine encodings.

import feedbackLexicon from "./data/feedback-lexicon.json";
import type { FeedbackEncoding, FeedbackInput, PreprocessResult } from "./types";

const URL_PATTERN = /https?:\/\/\S+|www\.\S+/gi;
const TAG_PATTERN = /@\w+/g;
const HASHTAG_PATTERN = /#\w+/g;

// Emoji pattern: matches unicode emoji sequences.
const EMOJI_PATTERN =
  /[\u{1F300}-\u{1F9FF}]|[\u{2600}-\u{27BF}]|[\u{1F000}-\u{1F02F}]|[\u{1F0A0}-\u{1F0FF}]|[\u{2300}-\u{23FF}]|[\u{2000}-\u{206F}]/gu;

// Pattern to detect letters repeated 3 or more times.
const REPETITION_PATTERN = /([a-zA-Z])\1{2,}/g;

const ABBREVIATION_MAP: Record<string, string> = {
  ...(feedbackLexicon.abbreviated_slang as Record<string, string>),
  ...(feedbackLexicon.abbreviated_cs_terms as Record<string, string>),
};

// Precomputed lowercase key → expansion, so each token is a single O(1) map hit.
const ABBREVIATION_LOOKUP = new Map(
  Object.entries(ABBREVIATION_MAP).map(([abbrev, fullWord]) => [abbrev.toLowerCase(), fullWord]),
);

// Strips URLs, @mentions, #hashtags, and emojis.
function removeNoise(text: string): string {
  let cleaned = text;

  cleaned = cleaned.replace(URL_PATTERN, "");
  cleaned = cleaned.replace(TAG_PATTERN, "");
  cleaned = cleaned.replace(HASHTAG_PATTERN, "");
  cleaned = cleaned.replace(EMOJI_PATTERN, "");

  return cleaned;
}

const seedVocabulary = new Set<string>(feedbackLexicon.seed_vocabulary as string[]);

function isValidWord(word: string): boolean {
  return seedVocabulary.has(word.toLowerCase());
}

// Reduces repeated letters (e.g., "yessss" → "yes") using vocabulary lookup.
function normalizeVowels(text: string): string {
  return text.replace(REPETITION_PATTERN, (match, letter, offset, fullString) => {
    const singleCharVersion = letter;
    const doubleCharVersion = letter + letter;

    let wordStart = offset;
    while (wordStart > 0 && /\w/.test(fullString[wordStart - 1])) {
      wordStart--;
    }
    let wordEnd = offset + match.length;
    while (wordEnd < fullString.length && /\w/.test(fullString[wordEnd])) {
      wordEnd++;
    }

    const originalWord = fullString.slice(wordStart, wordEnd);
    const doubleLetterWord = (
      originalWord.slice(0, offset - wordStart) +
      doubleCharVersion +
      originalWord.slice(offset + match.length - wordStart)
    ).toLowerCase();

    // All letters share one lookup path — "o"/"e" are not special-cased, so a
    // double-letter candidate must be a real word to be kept (e.g. "heey" → "hey").
    if (isValidWord(doubleLetterWord)) {
      return doubleCharVersion;
    }

    return singleCharVersion;
  });
}

// Splits a token into leading punctuation, the core word, and trailing
// punctuation, so abbreviation matching stays word-bounded.
function splitAttachedPunctuation(token: string): {
  core: string;
  leading: string;
  trailing: string;
} {
  let start = 0;
  while (start < token.length && !/\w/.test(token[start])) {
    start++;
  }
  let end = token.length;
  while (end > start && !/\w/.test(token[end - 1])) {
    end--;
  }
  return {
    core: token.slice(start, end),
    leading: token.slice(0, start),
    trailing: token.slice(end),
  };
}

// Expands abbreviations to full words in a single whitespace pass (O(n)).
function expandAbbreviations(text: string): string {
  return text
    .split(/(\s+)/)
    .map((chunk) => {
      if (/^\s+$/.test(chunk)) {
        return chunk;
      }

      const { core, leading, trailing } = splitAttachedPunctuation(chunk);
      const fullWord = ABBREVIATION_LOOKUP.get(core.toLowerCase());
      if (fullWord === undefined) {
        return chunk;
      }

      return leading + fullWord + trailing;
    })
    .join("");
}

function normalizeWhitespace(text: string): string {
  let normalized = text.replace(/\s+/g, " ");
  normalized = normalized.trim();

  return normalized;
}

// Fixed sequence length the fine-tuned ONNX models were trained with.
export const MAX_SEQ_LEN = 256;

// Structural tokenizer interface decoupling preprocessing from HF runtime imports.
export interface MachineTokenizer {
  (
    text: string,
    options?: Record<string, unknown>,
  ): {
    input_ids?: { data?: ArrayLike<number | bigint> };
    attention_mask?: { data?: ArrayLike<number | bigint> };
  };
}

// Normalizes feedback text (noise removal, vowel reduction, abbreviation mapping, whitespace).
export function CleanFeedback(feedback: FeedbackInput | string): string {
  const rawText = typeof feedback === "string" ? feedback : feedback.rawText;

  const afterNoise = removeNoise(rawText);
  const afterVowels = normalizeVowels(afterNoise);
  const afterAbbrevs = expandAbbreviations(afterVowels);
  return normalizeWhitespace(afterAbbrevs);
}

// Converts cleaned text into fixed-length BigInt64Array tensor encodings.
export function EncodeFeedback(
  cleanedText: string,
  tokenizer: MachineTokenizer,
  maxLength: number = MAX_SEQ_LEN,
): FeedbackEncoding {
  const output = tokenizer(cleanedText, {
    padding: "max_length",
    truncation: true,
    max_length: maxLength,
    return_tensor: true,
  });

  const inputIdsData = output?.input_ids?.data;
  const attentionMaskData = output?.attention_mask?.data;

  if (!inputIdsData || !attentionMaskData) {
    throw new Error("[preprocess] Tokenizer returned invalid encoding output.");
  }

  const toBigInt64 = (data: ArrayLike<number | bigint>): BigInt64Array =>
    data instanceof BigInt64Array ? data : BigInt64Array.from(data, (v) => BigInt(v));

  return {
    inputIds: toBigInt64(inputIdsData),
    attentionMask: toBigInt64(attentionMaskData),
  };
}

// Module 2: Preprocessing --- algorithm.pseudo line 9
export function Preprocess(feedback: FeedbackInput, tokenizer: MachineTokenizer): PreprocessResult {
  console.debug("[preprocess] INPUT BOUNDARY: Received feedback", {
    feedbackId: feedback.id,
    rawLength: feedback.rawText.length,
    sample: feedback.rawText.substring(0, 50),
  });

  const cleanedText = CleanFeedback(feedback);
  const encoding = EncodeFeedback(cleanedText, tokenizer);

  console.debug("[preprocess] OUTPUT BOUNDARY: Preprocessing complete", {
    cleanedLength: cleanedText.length,
    hasEncoding: !!encoding,
    sample: cleanedText.substring(0, 50),
  });

  return { cleanedText, encoding };
}

File: scripts/training/preprocess.py
"""
Preprocessing module (Python implementation of src/lib/algorithm/preprocess.ts)
Maintains 1:1 preprocessing parity between Python training loader and TS runtime adapter.
"""

import json
import re
from pathlib import Path

LEXICON_PATH = Path(__file__).resolve().parents[2] / "src" / "lib" / "algorithm" / "data" / "feedback-lexicon.json"

with LEXICON_PATH.open(encoding="utf-8") as handle:
    _LEXICON_DATA = json.load(handle)

# URL pattern: matches http://, https://, www., and domain-like patterns.
URL_PATTERN = re.compile(r"https?://\S+|www\.\S+", re.IGNORECASE)
# Mention/tag pattern: matches @username or similar tags.
TAG_PATTERN = re.compile(r"@\w+")
# Hashtag pattern: matches #hashtag.
HASHTAG_PATTERN = re.compile(r"#\w+")
# Emoji pattern: matches unicode emoji sequences.
EMOJI_PATTERN = re.compile(
    r"[\U0001F300-\U0001F9FF]|[\u2600-\u27BF]|[\U0001F000-\U0001F02F]|[\U0001F0A0-\U0001F0FF]|[\u2300-\u23FF]|[\u2000-\u206F]"
)

# Pattern to detect letters repeated 3 or more times.
REPETITION_PATTERN = re.compile(r"([a-zA-Z])\1{2,}")

# Extracted from common feedback patterns in educational contexts (English + Tagalog/Taglish).
ABBREVIATION_MAP = _LEXICON_DATA["abbreviations"]
ABBREVIATION_LOOKUP = {k.lower(): v for k, v in ABBREVIATION_MAP.items()}

# Double-letter vocabulary consulted by normalize_vowels to decide whether a
# repeated run collapses to one copy or keeps two.
SEED_VOCABULARY = set(_LEXICON_DATA["seed_vocabulary"])


def is_valid_word(word: str) -> bool:
    return word.lower() in SEED_VOCABULARY


def remove_noise(text: str) -> str:
    """Strips URLs, tags, hashtags, and emojis."""
    cleaned = URL_PATTERN.sub("", text)
    cleaned = TAG_PATTERN.sub("", cleaned)
    cleaned = HASHTAG_PATTERN.sub("", cleaned)
    cleaned = EMOJI_PATTERN.sub("", cleaned)
    return cleaned


def normalize_vowels(text: str) -> str:
    """Reduces repeated letters (e.g., 'yessss' -> 'yes') using vocabulary lookup."""

    def replace_match(match: re.Match) -> str:
        letter = match.group(1)
        offset = match.start()
        match_len = len(match.group(0))
        single_char_version = letter
        double_char_version = letter + letter

        # Find word boundaries around the match
        word_start = offset
        while word_start > 0 and (text[word_start - 1].isalnum() or text[word_start - 1] == "_"):
            word_start -= 1
        word_end = offset + match_len
        while word_end < len(text) and (text[word_end].isalnum() or text[word_end] == "_"):
            word_end += 1

        original_word = text[word_start:word_end]
        double_letter_word = (
            original_word[: offset - word_start]
            + double_char_version
            + original_word[offset + match_len - word_start :]
        ).lower()

        if is_valid_word(double_letter_word):
            return double_char_version

        return single_char_version

    return REPETITION_PATTERN.sub(replace_match, text)


def split_attached_punctuation(token: str):
    """Splits token into leading punctuation, core word, and trailing punctuation."""
    start = 0
    while start < len(token) and not (token[start].isalnum() or token[start] == "_"):
        start += 1
    end = len(token)
    while end > start and not (token[end - 1].isalnum() or token[end - 1] == "_"):
        end -= 1
    return token[:start], token[start:end], token[end:]


def expand_abbreviations(text: str) -> str:
    """Expands common abbreviations to full words using word-boundary single-pass tokenization."""
    chunks = re.split(r"(\s+)", text)
    result = []
    for chunk in chunks:
        if not chunk:
            continue
        if re.match(r"^\s+$", chunk):
            result.append(chunk)
            continue

        leading, core, trailing = split_attached_punctuation(chunk)
        full_word = ABBREVIATION_LOOKUP.get(core.lower())
        if full_word is None:
            result.append(chunk)
        else:
            result.append(leading + full_word + trailing)
    return "".join(result)


def normalize_whitespace(text: str) -> str:
    """Collapses whitespace and trims."""
    return re.sub(r"\s+", " ", text).strip()


def preprocess(text: str) -> str:
    """
    Cleans feedback text: removes noise, normalizes repetitions,
    expands abbreviations, trims whitespace.
    """
    if not isinstance(text, str):
        return ""
    after_noise = remove_noise(text)
    after_vowels = normalize_vowels(after_noise)
    after_abbrev = expand_abbreviations(after_vowels)
    cleaned = normalize_whitespace(after_abbrev)
    return cleaned
